csrc/scheduler/utils.cpp

```diff
@@ -2369,23 +2370,111 @@ std::vector<TensorView*> getResizedTensors(Fusion* fusion) {
   return resized_tensors;
 }
 
+// If the producer of a resized tensor is an input cache and we need to promote
+// it to global memory, just do not cache the input but directy read from the
+// global memory input. It doesn't make any sense to cache a global memory input
+// in global memory. Note that a copy of an input cache is inserted by
+// prepareForMemoryTypePromotion, so grab the producer of the
+// producer and see if it's included in input_caches.
+bool revertUseOfInputCacheInResize(
+    TensorView* resized_tensor,
+    TensorView* producer_of_resize,
+    MemoryType promoted_memory_type,
+    const std::vector<TensorView*>& input_caches) {
+  auto get_copy_src = [](TensorView* tv) -> TensorView* {
+    if (auto uop = dynamic_cast<UnaryOp*>(tv->definition())) {
+      if (uop->getUnaryOpType() == UnaryOpType::Set) {
+        return uop->in()->as<TensorView>();
+      }
+    }
+    return nullptr;
+  };
+
+  // Only applies if the promoted new type is Global
+  if (promoted_memory_type != MemoryType::Global) {
+    return true;
+  }
+
+  // To see if the producer is a cache of an input, need to look at
+  // its producer as a copy is inserted
+  auto producer_of_producer = get_copy_src(producer_of_resize);
+  if (producer_of_producer == nullptr) {
+    // No copy is detected. This must mean the producer is not a copy
+    // of an input cache
+    return false;
+  }
+
+  auto cache_it =
+      std::find(input_caches.begin(), input_caches.end(), producer_of_producer);
+  if (cache_it == input_caches.end()) {
+    return false;
+  }
+
+  auto fusion_input = get_copy_src(producer_of_producer);
+  TORCH_INTERNAL_ASSERT(
+      fusion_input != nullptr,
+      "Unexpected input cache: ",
+      producer_of_producer->toString());
+
+  // Currently, the ops look like:
+  // tv0: fusion input
+  // tv1 = tv0 // cache of the input
+  // tv2 = tv1 // copy of the tv1. Placed on Global
+  // tv3 = resizeOp(tv2) // some op using resize
+
+  //
+  // Translate it to:
+  // tv0: fusion input
+  // tv3 = resizeOp(tv0) // some op using resize
+
+  ir_utils::replaceValInExpr(
+      resized_tensor->definition(), producer_of_resize, fusion_input);
+
+  return true;
+}
+
 } // namespace
 
 void prepareForMemoryTypePromotion(Fusion* fusion) {
   auto resized_tensors = getResizedTensors(fusion);
   std::unordered_set<TensorView*> cached;
   for (auto resized_tensor : resized_tensors) {
-    for (auto producer : ir_utils::producerTvsOf(resized_tensor)) {
-      if (cached.count(producer) != 0) {
-        continue;
-      }
-      producer->cacheAfter();
-      cached.insert(producer);
+    // Resized tensors are those created by operations like pad and
```

Comment:
This is an unrelated change. Instead of using `cacheAfter`, just create a copy of `producer` and use it as the input to the resize expr. 

Example:

```
tv1 = sum(tv0)
tv2 = some_resize_op(tv1);
tv3 = some_other_op(tv1);
```

When `tv1` is promoted to `Global`, we want to avoid reducing to a global memory tensor, so with `cacheAfter`:

```
tv1 = sum(tv0);
tv4 = tv1
tv4->setMemoryType(Global)
tv2 = some_resize_op(tv3)
tv3 = some_other_op(tv3);
```

This way, the reduction is done using Local, but `some_other_op` doesn't need to use the gmem copy of `tv1`. This should be just fine:

```
tv1 = sum(tv0);
tv4 = tv1
tv4->setMemoryType(Global)
tv2 = some_resize_op(tv4)
tv3 = some_other_op(tv1);
```

Reply:
Is this just `cacheFork`?

Reply:
I think we can extend the interface of `cacheFork` to do this?

If we have:
```C++
tv1 = sum(tv0)
tv2 = some_resize_op(tv1);
tv3 = some_other_op(tv1);
```

Can we just do 
```C++
tv1->setMemoryType(Global);
tv4 = tv1->cacheFork(/*keep_global=*/{tv2->definition()})
```

And we get:
```
tv0 -> tv4 (local) -> tv1 -> tv2
       tv4 (fork) -> tv3
```

Just some idea, have no strong opinion on having to do like this.

Reply:
`cacheFork` is designed to change Fusion outputs. We could make it optional, of course. This is a simple transformation, so I don't feel it's worth consolidating into an extended `cacheFork`